src/sinks/util/http.rs

```diff
@@ -598,6 +599,145 @@ pub fn validate_headers(
     Ok(validated_headers)
 }
 
+/// Request type for use in the `Service` implementation of HTTP stream sinks.
+#[derive(Clone)]
+pub struct HttpRequest {
+    payload: Bytes,
+    finalizers: EventFinalizers,
+    request_metadata: RequestMetadata,
+}
+
+impl HttpRequest {
+    /// Creates a new `HttpRequest`.
+    pub fn new(
+        payload: Bytes,
+        finalizers: EventFinalizers,
+        request_metadata: RequestMetadata,
+    ) -> Self {
+        Self {
+            payload,
+            finalizers,
+            request_metadata,
+        }
+    }
+}
+
+impl Finalizable for HttpRequest {
+    fn take_finalizers(&mut self) -> EventFinalizers {
+        self.finalizers.take_finalizers()
+    }
+}
+
+impl MetaDescriptive for HttpRequest {
+    fn get_metadata(&self) -> &RequestMetadata {
+        &self.request_metadata
+    }
+
+    fn metadata_mut(&mut self) -> &mut RequestMetadata {
+        &mut self.request_metadata
+    }
+}
+
+impl ByteSizeOf for HttpRequest {
+    fn allocated_bytes(&self) -> usize {
+        self.payload.allocated_bytes() + self.finalizers.allocated_bytes()
+    }
+}
+
+/// Response type for use in the `Service` implementation of HTTP stream sinks.
+pub struct HttpResponse {
+    pub http_response: Response<Bytes>,
+    events_byte_size: GroupedCountByteSize,
+    raw_byte_size: usize,
+}
+
+impl DriverResponse for HttpResponse {
+    fn event_status(&self) -> EventStatus {
+        if self.http_response.status().is_success() {
+            EventStatus::Delivered
+        } else {
+            EventStatus::Rejected
+        }
```

Comment:
I think we should prefer to use the `super::sink::Response` trait here to easily track transient responses as well.

```suggestion
        if self.http_response.is_successful() {
            EventStatus::Delivered
        } else if self.http_response.is_transient() {
            EventStatus::Errored
        } else {
            EventStatus::Rejected
        }
```

Reply:
That seems like the right thing to do here, but it's worth noting that it's introducing a behavior change.